File: src/bus.rs
use crate::{
    cartridge::Cartridge,
    memory::{Error, Read, Write},
    vram::Vram,
};

pub struct Bus {
    interrupt_enable_reg: u8,
    high_ram: [u8; 0x007F],
    vram: Vram,
    cartridge: Cartridge,
}

impl Bus {
    pub fn new(cartridge: Cartridge) -> Self {
        Self {
            interrupt_enable_reg: 0,
            high_ram: [0; 0x007F],
            vram: Vram::new(cartridge.cgb()),
            cartridge,
        }
    }
}

impl Read for Bus {
    fn read(&self, address: u16) -> Result<u8, Error> {
        match address {
            0x0000..0x8000 => self.cartridge.rom().read(address),
            0x8000..0xA000 => self.vram.read(address.saturating_sub(0x8000)),
            0xA000..0xC000 => self.cartridge.ram().read(address.saturating_sub(0xA000)),
            0xC000..0xE000 => todo!("Wram"),
            0xE000..0xFE00 => todo!("Echo ram"),
            0xFE00..0xFEA0 => todo!("Object attribute memory"),
            0xFEA0..0xFF00 => Err(Error::SegFault { address }), // Prohibited usage -> segfault
            0xFF00..0xFF80 => todo!("IO registers"),
            0xFF80..0xFFFF => self.high_ram.read(address.saturating_sub(0xFF80)),
            0xFFFF => Ok(self.interrupt_enable_reg),
        }
        .map_err(|e| match e {
            Error::SegFault { address: _ } => Error::SegFault { address },
        })
    }
}

impl Write for Bus {
    fn write(&mut self, address: u16, value: u8) -> Result<(), Error> {
        match address {
            0x0000..0x8000 => self.cartridge.rom_mut().write(address, value),
            0x8000..0xA000 => self.vram.write(address.saturating_sub(0x8000), value),
            0xA000..0xC000 => self
                .cartridge
                .ram_mut()
                .write(address.saturating_sub(0xA000), value),
            0xC000..0xE000 => todo!("Wram"),
            0xE000..0xFE00 => Err(Error::SegFault { address }), // prohibited usage of echo ram -> segfault
            0xFE00..0xFEA0 => todo!("Object attribute memory"),
            0xFEA0..0xFF00 => Err(Error::SegFault { address }), // Prohibited usage -> segfault
            0xFF00..0xFF80 => todo!("IO registers"),
            0xFF80..0xFFFF => self.high_ram.write(address.saturating_sub(0xFF80), value),
            0xFFFF => {
                self.interrupt_enable_reg = value;
                Ok(())
            }
        }
        .map_err(|e| match e {
            Error::SegFault { address: _ } => Error::SegFault { address },
        })
    }
}

File: Boii/IO/Bus.cs
using Boii.Abstractions;
using Boii.Errors;
using Boii.Memory;

namespace Boii.IO;

public class Bus : IGenericIO
{
    private readonly record struct Range(ushort Lower, ushort Upper)
    {
        public bool IsIn(ushort address, out ushort relativeAddress)
        {
            relativeAddress = (ushort)(address - Lower);
            return Lower <= address && address < Upper;
        }
    }

    private static readonly Range CartridgeRomRange = new(0x0000, 0x8000);
    private static readonly Range VramRange = new(0x8000, 0xA000);
    private static readonly Range CartridgeRamRange = new(0xA000, 0xC000);
    private static readonly Range WorkRamRange = new(0xC000, 0xE000);
    private static readonly Range EchoRamRange = new(0xE000, 0xFE00);
    private static readonly Range ObjectAttributeMemoryRange = new(0xFE00, 0xFEA0);
    private static readonly Range IoRegistersRange = new(0xFF00, 0xFF80);
    private static readonly Range HighRamRange = new(0xFF80, 0xFFFF);

    public IGenericIO? CartridgeRom { get; set; }
    public IGenericIO? VideoRam { get; set; }
    public IGenericIO? CartridgeRam { get; set; }
    private readonly ArrayMemory _workRam = ArrayMemory.Create("WRAM", 0x2000);
    public IGenericIO? ObjectAttributeMemory { get; set; }
    public IGenericIO? IoRegisters { get; set; }
    private readonly ArrayMemory _highRam = ArrayMemory.Create("HRAM", 0x007F);
    private byte _interruptEnable = 0;

    private Bus() { }

    public static Bus CreateWithoutLinks() => new();

    public byte Read(ushort address) => address switch
    {
        var x when CartridgeRomRange.IsIn(x, out var i) && CartridgeRom is not null => CartridgeRom.Read(i),
        var x when VramRange.IsIn(x, out var i) && VideoRam is not null => VideoRam.Read(i),
        var x when CartridgeRamRange.IsIn(x, out var i) && CartridgeRam is not null => CartridgeRam.Read(i),
        var x when WorkRamRange.IsIn(x, out var i) => _workRam.Read(i),
        var x when EchoRamRange.IsIn(x, out var i) => _workRam.Read(i),     // Mirrors wram
        var x when ObjectAttributeMemoryRange.IsIn(x, out var i) && ObjectAttributeMemory is not null => ObjectAttributeMemory.Read(i),
        var x when IoRegistersRange.IsIn(x, out var i) && IoRegisters is not null => IoRegisters.Read(i),
        var x when HighRamRange.IsIn(x, out var i) => _highRam.Read(i),
        0xFFFF => _interruptEnable,
        _ => throw SegmentationFault.Create($"{nameof(Bus)}.{nameof(Read)}", address),
    };

    public void Write(ushort address, byte value)
    {
        if (CartridgeRomRange.IsIn(address, out var x) && CartridgeRom is not null)
            CartridgeRom.Write(x, value);
        else if (VramRange.IsIn(address, out x) && VideoRam is not null)
            VideoRam.Write(x, value);
        else if (CartridgeRamRange.IsIn(address, out x) && CartridgeRam is not null)
            CartridgeRam.Write(x, value);
        else if (WorkRamRange.IsIn(address, out x))
            _workRam.Write(x, value);
        else if (EchoRamRange.IsIn(address, out x))
            _workRam.Write(x, value);
        else if (ObjectAttributeMemoryRange.IsIn(address, out x) && ObjectAttributeMemory is not null)
            ObjectAttributeMemory.Write(x, value);
        else if (IoRegistersRange.IsIn(address, out x) && IoRegisters is not null)
            IoRegisters.Write(x, value);
        else if (HighRamRange.IsIn(address, out x))
            _highRam.Write(x, value);
        else if (address == 0xFFFF)
            _interruptEnable = value;
        else
            throw SegmentationFault.Create($"{nameof(Bus)}.{nameof(Write)}", address);
    }
}
